### PR TITLE
cloud: add read and write metrics

### LINKED ISSUE #89242: cloud: add a metrics recorder that tracks per provider metrics
There have been numerous support issues where we are unable to conclusively determine whether reading/writing from ExternalStorage is our bottleneck when investigating a slow backup/restore. In 22.2 we have improved our traces to be able to better observe how much time we are spending in a `WriteFile` or `ReadFile` method call over the lifetime of a job, but collecting these traces/visualizing them are still non-trivial. While we continue improving the tracing experience we have long discussed collecting simple ExternalStorage provider metrics that we can surface on our Debug Console Dashboards. Following is a non-exhaustive list of metrics that are easy to collect and display, and will provide a valuable signal that helps answer where our backup/restore is slow.

- [ ] `cloud.write`
- [ ] `cloud.read`
- [ ] `cloud.retries`
- [ ] `cloud.errors`

Note, this would mean that interactions with all external storage providers would contribute to the same metric, but this is still useful and better than no signal. Also more often than not a single cluster is only interacting with a single cloud provider if we feel the need for granularity we could explore using provider/bucket specific labels that can be visualized in prometheus. When investigating a slow cluster backup or restore we often have periods of time where the job is the only thing running and so it will be easy to correlate the bytes read/written to the running operations in a cluster.

We already have the infrastructure to wrap our readers/writers for the purposes of rate limiting/accounting and so this project will probably involve sticking a metrics recorder on the `esWrapper` and ensuring every read and write call is accounted for. 

Jira issue: CRDB-20147

Epic: CRDB-20265

### PR DESCRIPTION
**Pull request by adityamaru**
This change adds cloud.read_bytes and cloud.write_bytes metrics.
These are updated on all read/write operations to external storage endpoints.

These metrics while shared by all external storage interactions
will provide an immediate signal into whether we're seeing an
abnormal rate of reads and writes errors during support investigations.

Fixes: https://github.com/cockroachdb/cockroach/issues/89242

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
A 2TB import; I paused it in the middle to fiddle with some cluster settings for funsies.

Screen Shot 2022-10-07 at 2 55 06 PM

And then a backup of that data:

Screen Shot 2022-10-07 at 2 55 18 PM

**Review by dt — COMMENTED** [review]
Are we concerned that the retry count is potentially not going to reflect retries in the implementation of some clients and/or at a higher layer either? A misleading metric might be worse than no metric eg if someone looks at it, sees it is flat and dismisses retries even if they're happening inside foo-sdk or library?

**Comment by adityamaru** [comment]
> Are we concerned that the retry count is potentially not going to reflect retries in the implementation of some clients and/or at a higher layer either?

Yeah that's a good point and one that crossed my mind when I was plumbing counter increments into all the retry loops. We could try to be diligent about having all clients read and write paths that we retry in `pkg/cloud/<provider>`  increment this counter but maybe that's too optimistic. This doesn't address internal SDK retries or higher-level job retries either but maybe tweaking the metric description to make this clear solves that?

Alternatively, we could breakdown the retries into `cloud.resumingreader.error.retries`, `cloud.custom.error.retries` to better define what they track? The latter being any custom error handling we inject into the SDKs.

**Review by dt — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!

bors r=dt

File: pkg/roachpb/api.proto [hunk 1]
```diff
@@ -1464,13 +1464,6 @@ message ComputeChecksumResponse {
       (gogoproto.customtype) = "github.com/cockroachdb/cockroach/pkg/util/uuid.UUID"];
 }
 
-// RetryTracingEvent is the trace recording used to track retries.
-message RetryTracingEvent {
-  string operation = 1;
-  int32 attempt_number = 2;
-  string retry_error = 3;
-}
```
Comment (stevendanna) [inline pkg/roachpb/api.proto:1472]: I was concerned that it wasn't safe to remove this message type in the case of importing a remote recording from a node on an older version, but it looks to me like we ignore unmarshalling errors, at least when sending them to listeners.

Comment (adityamaru) [reply pkg/roachpb/api.proto:1472]: Very good point but yeah https://github.com/cockroachdb/cockroach/blob/2d291751692bb40cf634504c74235ff25d1c80aa/pkg/util/tracing/crdbspan.go#L814 looks like we don't error out if unmarshalling failed. I could keep the proto around and just stop sending it for a release out of caution.

Comment (adityamaru) [reply pkg/roachpb/api.proto]: added the proto back, will delete it in 23.2

**Comment by adityamaru** [comment]
bors r=dt,stevendanna

**Comment by adityamaru** [comment]
bors retry

**Pull request merged**